icepyx/core/read.py

```diff
@@ -0,0 +1,616 @@
+                #                     "description": "Beam Number",
+                #                     "type": "str",
+                #                     "default": "gt1l",
+                #                     "allowed": ["gt1l", "gt1r", "gt2l", "gt2r", "gt3l", "gt3r"]
+                #                 }],
+            )
+            ds = grpcat[self._source_type].read()
+
+        # NOTE: could also do this with h5py, but then would have to read in each variable in the group separately
+        except ValueError:
+            grpcat = is2cat.build_catalog(
+                file,
+                self._pattern,
+                self._source_type,
+                grp_paths=grp_path,
+                extra_engine_kwargs={"phony_dims": "access"},
+            )
+
+            ds = grpcat[self._source_type].read()
+
+        return ds
+
+    # NOTE: for non-gridded datasets only
+    def _build_single_file_dataset(self, file, groups_list):
+        """
+        Create a single xarray dataset with all of the wanted variables/groups from the wanted var list for a single data file/url.
+
+        Parameters
+        ----------
+        file : str
+            Full path to ICESat-2 data file.
+            Currently tested for locally downloaded files; untested but hopefully works for s3 stored cloud files.
+
+        groups_list : list of strings
+            List of full paths to data variables within the file.
+            e.g. ['orbit_info/sc_orient', 'gt1l/land_ice_segments/h_li', 'gt1l/land_ice_segments/latitude', 'gt1l/land_ice_segments/longitude']
+
+        Returns
+        -------
+        Xarray Dataset
+        """
+
+        file_product = self._read_single_var(file, "/").attrs["identifier_product_type"]
+        assert (
+            file_product == self._prod
+        ), "Your product specification does not match the product specification within your files."
+        # I think the below method might NOT read the file into memory as the above might?
+        # import h5py
+
+        # with h5py.File(filepath,'r') as h5pt:
+        #     prod_id = h5pt.attrs["identifier_product_type"]
+
+        is2ds = self._build_dataset_template(file)
+
+        # returns the wanted groups as a single list of full group path strings
+        wanted_dict, wanted_groups = Variables.parse_var_list(groups_list, tiered=False)
+        wanted_groups_set = set(wanted_groups)
+        # orbit_info is used automatically as the first group path so the info is available for the rest of the groups
+        wanted_groups_set.remove("orbit_info")
+        # returns the wanted groups as a list of lists with group path string elements separated
+        _, wanted_groups_tiered = Variables.parse_var_list(groups_list, tiered=True)
+
+        for grp_path in ["orbit_info"] + list(wanted_groups_set):
+            # print(grp_path)
+            ds = self._read_single_var(file, grp_path)
+            is2ds = Read._add_var_to_ds(
```

Comment:
If I'm remembering correctly, using `self._add_var_to_ds()` here caused an error because `_add_var_to_ds()` is a [static method](https://www.geeksforgeeks.org/class-method-vs-static-method-python/), so it thinks it's getting too many inputs.